// reductoai/index.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var _Reducto_instances, _a, _Reducto_baseURLOverridden;
import { isRequestOptions } from "./core.mjs";
import * as Core from "./core.mjs";
import * as Errors from "./error.mjs";
import * as Uploads from "./uploads.mjs";
import * as API from "./resources/index.mjs";
import { Config } from "./resources/config.mjs";
import { Edit } from "./resources/edit.mjs";
import { Extract, } from "./resources/extract.mjs";
import { Job } from "./resources/job.mjs";
import { Parse, } from "./resources/parse.mjs";
import { Pipeline, } from "./resources/pipeline.mjs";
import { Split } from "./resources/split.mjs";
import { Webhook } from "./resources/webhook.mjs";
const environments = {
    production: 'https://platform.reducto.ai',
    eu: 'https://eu.platform.reducto.ai',
    au: 'https://au.platform.reducto.ai',
};
/**
 * API Client for interfacing with the Reducto API.
 */
export class Reducto extends Core.APIClient {
    /**
     * API Client for interfacing with the Reducto API.
     *
     * @param {string | undefined} [opts.apiKey=process.env['REDUCTO_API_KEY'] ?? undefined]
     * @param {Environment} [opts.environment=production] - Specifies the environment URL to use for the API.
     * @param {string} [opts.baseURL=process.env['REDUCTO_BASE_URL'] ?? https://platform.reducto.ai] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 hour] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = Core.readEnv('REDUCTO_BASE_URL'), apiKey = Core.readEnv('REDUCTO_API_KEY'), ...opts } = {}) {
        if (apiKey === undefined) {
            throw new Errors.ReductoError("The REDUCTO_API_KEY environment variable is missing or empty; either provide it, or instantiate the Reducto client with an apiKey option, like new Reducto({ apiKey: 'My API Key' }).");
        }
        const options = {
            apiKey,
            ...opts,
            baseURL,
            environment: opts.environment ?? 'production',
        };
        if (baseURL && opts.environment) {
            throw new Errors.ReductoError('Ambiguous URL; The `baseURL` option (or REDUCTO_BASE_URL env var) and the `environment` option are given. If you want to use the environment you must pass baseURL: null');
        }
        super({
            baseURL: options.baseURL || environments[options.environment || 'production'],
            baseURLOverridden: baseURL ? baseURL !== environments[options.environment || 'production'] : false,
            timeout: options.timeout ?? 3600000 /* 1 hour */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        _Reducto_instances.add(this);
        this.job = new API.Job(this);
        this.split = new API.Split(this);
        this.parse = new API.Parse(this);
        this.extract = new API.Extract(this);
        this.edit = new API.Edit(this);
        this.pipeline = new API.Pipeline(this);
        this.webhook = new API.Webhook(this);
        this.config = new API.Config(this);
        this._options = options;
        this.apiKey = apiKey;
    }
    /**
     * Get Version
     */
    apiVersion(options) {
        return this.get('/version', options);
    }
    upload(params = {}, options) {
        if (isRequestOptions(params)) {
            return this.upload({}, params);
        }
        const { extension, ...body } = params;
        return this.post('/upload', Core.maybeMultipartFormRequestOptions({ query: { extension }, body, ...options }));
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { Authorization: `Bearer ${this.apiKey}` };
    }
}
_a = Reducto, _Reducto_instances = new WeakSet(), _Reducto_baseURLOverridden = function _Reducto_baseURLOverridden() {
    return this.baseURL !== environments[this._options.environment || 'production'];
};
Reducto.Reducto = _a;
Reducto.DEFAULT_TIMEOUT = 3600000; // 1 hour
Reducto.ReductoError = Errors.ReductoError;
Reducto.APIError = Errors.APIError;
Reducto.APIConnectionError = Errors.APIConnectionError;
Reducto.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Reducto.APIUserAbortError = Errors.APIUserAbortError;
Reducto.NotFoundError = Errors.NotFoundError;
Reducto.ConflictError = Errors.ConflictError;
Reducto.RateLimitError = Errors.RateLimitError;
Reducto.BadRequestError = Errors.BadRequestError;
Reducto.AuthenticationError = Errors.AuthenticationError;
Reducto.InternalServerError = Errors.InternalServerError;
Reducto.PermissionDeniedError = Errors.PermissionDeniedError;
Reducto.UnprocessableEntityError = Errors.UnprocessableEntityError;
Reducto.toFile = Uploads.toFile;
Reducto.fileFromPath = Uploads.fileFromPath;
Reducto.Job = Job;
Reducto.Split = Split;
Reducto.Parse = Parse;
Reducto.Extract = Extract;
Reducto.Edit = Edit;
Reducto.Pipeline = Pipeline;
Reducto.Webhook = Webhook;
Reducto.Config = Config;
export { toFile, fileFromPath } from "./uploads.mjs";
export { ReductoError, APIError, APIConnectionError, APIConnectionTimeoutError, APIUserAbortError, NotFoundError, ConflictError, RateLimitError, BadRequestError, AuthenticationError, InternalServerError, PermissionDeniedError, UnprocessableEntityError, } from "./error.mjs";
export default Reducto;
