// reductoai/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _Reducto_instances, _a, _Reducto_baseURLOverridden;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnprocessableEntityError = exports.PermissionDeniedError = exports.InternalServerError = exports.AuthenticationError = exports.BadRequestError = exports.RateLimitError = exports.ConflictError = exports.NotFoundError = exports.APIUserAbortError = exports.APIConnectionTimeoutError = exports.APIConnectionError = exports.APIError = exports.ReductoError = exports.fileFromPath = exports.toFile = exports.Reducto = void 0;
const core_1 = require("./core.js");
const Core = __importStar(require("./core.js"));
const Errors = __importStar(require("./error.js"));
const Uploads = __importStar(require("./uploads.js"));
const API = __importStar(require("./resources/index.js"));
const config_1 = require("./resources/config.js");
const edit_1 = require("./resources/edit.js");
const extract_1 = require("./resources/extract.js");
const job_1 = require("./resources/job.js");
const parse_1 = require("./resources/parse.js");
const pipeline_1 = require("./resources/pipeline.js");
const split_1 = require("./resources/split.js");
const webhook_1 = require("./resources/webhook.js");
const environments = {
    production: 'https://platform.reducto.ai',
    eu: 'https://eu.platform.reducto.ai',
    au: 'https://au.platform.reducto.ai',
};
/**
 * API Client for interfacing with the Reducto API.
 */
class Reducto extends Core.APIClient {
    /**
     * API Client for interfacing with the Reducto API.
     *
     * @param {string | undefined} [opts.apiKey=process.env['REDUCTO_API_KEY'] ?? undefined]
     * @param {Environment} [opts.environment=production] - Specifies the environment URL to use for the API.
     * @param {string} [opts.baseURL=process.env['REDUCTO_BASE_URL'] ?? https://platform.reducto.ai] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 hour] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = Core.readEnv('REDUCTO_BASE_URL'), apiKey = Core.readEnv('REDUCTO_API_KEY'), ...opts } = {}) {
        if (apiKey === undefined) {
            throw new Errors.ReductoError("The REDUCTO_API_KEY environment variable is missing or empty; either provide it, or instantiate the Reducto client with an apiKey option, like new Reducto({ apiKey: 'My API Key' }).");
        }
        const options = {
            apiKey,
            ...opts,
            baseURL,
            environment: opts.environment ?? 'production',
        };
        if (baseURL && opts.environment) {
            throw new Errors.ReductoError('Ambiguous URL; The `baseURL` option (or REDUCTO_BASE_URL env var) and the `environment` option are given. If you want to use the environment you must pass baseURL: null');
        }
        super({
            baseURL: options.baseURL || environments[options.environment || 'production'],
            baseURLOverridden: baseURL ? baseURL !== environments[options.environment || 'production'] : false,
            timeout: options.timeout ?? 3600000 /* 1 hour */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        _Reducto_instances.add(this);
        this.job = new API.Job(this);
        this.split = new API.Split(this);
        this.parse = new API.Parse(this);
        this.extract = new API.Extract(this);
        this.edit = new API.Edit(this);
        this.pipeline = new API.Pipeline(this);
        this.webhook = new API.Webhook(this);
        this.config = new API.Config(this);
        this._options = options;
        this.apiKey = apiKey;
    }
    /**
     * Get Version
     */
    apiVersion(options) {
        return this.get('/version', options);
    }
    upload(params = {}, options) {
        if ((0, core_1.isRequestOptions)(params)) {
            return this.upload({}, params);
        }
        const { extension, ...body } = params;
        return this.post('/upload', Core.maybeMultipartFormRequestOptions({ query: { extension }, body, ...options }));
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { Authorization: `Bearer ${this.apiKey}` };
    }
}
exports.Reducto = Reducto;
_a = Reducto, _Reducto_instances = new WeakSet(), _Reducto_baseURLOverridden = function _Reducto_baseURLOverridden() {
    return this.baseURL !== environments[this._options.environment || 'production'];
};
Reducto.Reducto = _a;
Reducto.DEFAULT_TIMEOUT = 3600000; // 1 hour
Reducto.ReductoError = Errors.ReductoError;
Reducto.APIError = Errors.APIError;
Reducto.APIConnectionError = Errors.APIConnectionError;
Reducto.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Reducto.APIUserAbortError = Errors.APIUserAbortError;
Reducto.NotFoundError = Errors.NotFoundError;
Reducto.ConflictError = Errors.ConflictError;
Reducto.RateLimitError = Errors.RateLimitError;
Reducto.BadRequestError = Errors.BadRequestError;
Reducto.AuthenticationError = Errors.AuthenticationError;
Reducto.InternalServerError = Errors.InternalServerError;
Reducto.PermissionDeniedError = Errors.PermissionDeniedError;
Reducto.UnprocessableEntityError = Errors.UnprocessableEntityError;
Reducto.toFile = Uploads.toFile;
Reducto.fileFromPath = Uploads.fileFromPath;
Reducto.Job = job_1.Job;
Reducto.Split = split_1.Split;
Reducto.Parse = parse_1.Parse;
Reducto.Extract = extract_1.Extract;
Reducto.Edit = edit_1.Edit;
Reducto.Pipeline = pipeline_1.Pipeline;
Reducto.Webhook = webhook_1.Webhook;
Reducto.Config = config_1.Config;
var uploads_1 = require("./uploads.js");
Object.defineProperty(exports, "toFile", { enumerable: true, get: function () { return uploads_1.toFile; } });
Object.defineProperty(exports, "fileFromPath", { enumerable: true, get: function () { return uploads_1.fileFromPath; } });
var error_1 = require("./error.js");
Object.defineProperty(exports, "ReductoError", { enumerable: true, get: function () { return error_1.ReductoError; } });
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return error_1.APIError; } });
Object.defineProperty(exports, "APIConnectionError", { enumerable: true, get: function () { return error_1.APIConnectionError; } });
Object.defineProperty(exports, "APIConnectionTimeoutError", { enumerable: true, get: function () { return error_1.APIConnectionTimeoutError; } });
Object.defineProperty(exports, "APIUserAbortError", { enumerable: true, get: function () { return error_1.APIUserAbortError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return error_1.NotFoundError; } });
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return error_1.ConflictError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return error_1.RateLimitError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return error_1.BadRequestError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return error_1.AuthenticationError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return error_1.InternalServerError; } });
Object.defineProperty(exports, "PermissionDeniedError", { enumerable: true, get: function () { return error_1.PermissionDeniedError; } });
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return error_1.UnprocessableEntityError; } });
exports = module.exports = Reducto;
exports.default = Reducto;
